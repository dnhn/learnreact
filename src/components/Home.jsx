import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import {
  getPhotos,
  openAbout,
  requestPhotos,
} from '../redux/actions';

import './Home.scss';

import About from './About';
import PhotoThumbnail from './PhotoThumbnail';
import PhotoPopup from './PhotoPopup';
import Loading from './Loading';
import GitHubMark from './GitHubMark';

export const Home =
  ({
    aboutVisibility,
    photos,
    selectedPhoto,
    openAbout,
    requestPhotos,
    getPhotos,
  }) => {
  useEffect(() => {
    requestPhotos();
    getPhotos();
  }, []);

  return (
    <main>
      <header>
        <nav className="nav">
          <button
            type="button"
            onClick={openAbout}
          >
            About
          </button>
        </nav>
      </header>
      <GitHubMark />
      <section className="photos">
        {photos.requesting ?
          <Loading /> :
          photos.error ?
            <Loading error /> :
            photos.list.map(
              (p, idx) =>
                <PhotoThumbnail
                  key={p.id}
                  data={p}
                  order={idx}
                />
            )}
      </section>
      {selectedPhoto && <PhotoPopup />}
      {aboutVisibility && <About />}
    </main>
  );
}

const mapStateToProps = state => ({
  photos: state.photos,
  selectedPhoto: state.selectedPhoto,
  aboutVisibility: state.aboutVisibility,
});

const mapDispatchToProps = dispatch => ({
  openAbout: () => dispatch(openAbout()),
  requestPhotos: () => dispatch(requestPhotos()),
  getPhotos: () => dispatch(getPhotos()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);
