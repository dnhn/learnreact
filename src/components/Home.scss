*,
*::before,
*::after {
  box-sizing: border-box;
}

html,
body,
#root,
main,
.nav {
  height: 100%;
}

body {
  margin: 0;
  padding: 0;
  background-image: radial-gradient(#666, black);
  background-color: black;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

*:focus {
  outline: 4px dashed red;
}

main {
  flex-direction: column;
}

.nav button {
  cursor: pointer;
  border: 0;
  height: 100%;
  background-color: #666;
  color: white;
  font-weight: bold;

  &:hover,
  &:focus,
  &:active {
    background-color: #333;
  }
}

header {
  position: relative;
  overflow: visible;
  z-index: 5;
  width: 100%;
  height: 6%;
  background-color: white;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 2%;
}

.photos {
  flex: 1;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-content: center;
}

@media screen and (max-width: 640px) {
  body {
    align-items: flex-start;
  }

  .photos {
    height: 85%;
  }
}

@media screen and (min-width: 641px) {
  main {
    display: flex;
  }

  .nav button {
    padding: 10px;
  }

  header {
    height: 5%;
  }
}
