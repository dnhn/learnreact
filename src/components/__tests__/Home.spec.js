import React from 'react';
import { fireEvent, render, waitForDomChange } from '@testing-library/react';

import { Home } from '../Home';

let mockStore = {};

beforeEach(() => {
  mockStore = {
    selectedPhoto: null,
    photos: {
      requesting: false,
      error: false,
      list: [],
    },
    requestPhotos: jest.fn(),
    getPhotos: jest.fn(),
  };
});

describe('<Home />', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Home {...mockStore} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('makes requests', () => {
    render(<Home {...mockStore} />);
    expect(mockStore.requestPhotos).toHaveBeenCalled();
    expect(mockStore.getPhotos).toHaveBeenCalled();
  });

  it('loading photos', () => {
    mockStore.photos.requesting = true;
    const { baseElement } = render(<Home {...mockStore} />);
    expect(
      baseElement
        .getElementsByClassName('loading')
        .length
    )
      .toBe(1);
  });

  it('failed loading photos', () => {
    mockStore.photos.error = true;
    const { baseElement } = render(<Home {...mockStore} />);
    expect(
      baseElement
        .getElementsByClassName('loading')[0]
        .classList
        .contains('error')
    )
      .toBe(true);
  });

  it('opens about popup', async () => {
    const { container, getByText } = render(<Home {...mockStore} />);
    fireEvent.click(getByText('About'));
    waitForDomChange(container).then(
      () => expect(container.getElementsByClassName('about').length)
        .toBe(1)
    );
  });
});
