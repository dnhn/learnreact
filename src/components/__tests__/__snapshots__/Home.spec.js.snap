// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders without crashing 1`] = `
<DocumentFragment>
  <main>
    <header>
      <nav
        class="nav"
      >
        <button
          type="button"
        >
          About
        </button>
      </nav>
    </header>
    <a
      class="logo"
      href="https://github.com/nhantdn/learnreact"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="Hosted by GitHub"
        src="github-mark-light.svg"
      />
    </a>
    <section
      class="photos"
    />
  </main>
</DocumentFragment>
`;
